torch/optim/optimizer.py

```diff
@@ -54,6 +54,19 @@ def _dispatch_sqrt(x: float):  # float annotation is needed because of torchscri
     else:
         return math.sqrt(x)
 
+
+# We try to use the foreach implementation on CUDA whenever possible since
+# it is faster than the for-loop implementation. However, the foreach
+# implementation is not differentiable, so we must check differentiable=False.
+def _default_to_foreach(tensorlists: List[List[torch.Tensor]], differentiable: bool = False) -> bool:
+    all_tensors = []
+    for tensorlist in tensorlists:
+        all_tensors.extend(tensorlist)
+    return not torch.jit.is_scripting() and not differentiable and all(
```

Comment:
These booleans have nothing to do with iterating over the list, can we have it be an early return before the all_tensor list is even constructed?

Reply:
On 2nd thought, this isn't perf critical so fine as is.

Reply:
haha appreciate the perf-oriented mindset! since this is so close to landing, i'll leave it for now but will update in a later PR.